core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -97,6 +97,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case requestId => throw new KafkaException("Unknown api code " + requestId)
       }
     } catch {
+      case fee: FatalExitError => throw fee
```

Comment:
Thanks @ijuma for taking over and sorry for not being able to spend more time on this. My approach was i) to exclude FatalExitError from all catch clauses that is too broad, and ii) to instrument all threads to be FatalExitError-aware. I looked for broad catch clauses by grepping Throwable and apparently missed this catch clause (or perhaps some more?) that do not specify the exception type: https://github.com/apache/kafka/blob/128d0ff91d84a3a1f5a5237133f9ec01caf18d66/core/src/main/scala/kafka/server/KafkaRequestHandler.scala#L63

Regarding the first part of your question I do not remember if there was a particular Exception that i was expecting to reach apis but my approach was to have the entire code base consistent so it would be resilient against any missed existing case or any case that might be introduced by future changes.